src/iop/hazeremoval.c

```diff
@@ -0,0 +1,962 @@
+      Notes in Computer Science, pp. 1-14, 2010. DOI:
+      10.1007/978-3-642-15549-9_1
+*/
+
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "common/darktable.h"
+#include "bauhaus/bauhaus.h"
+#include "develop/imageop.h"
+#include "develop/tiling.h"
+#include "gui/gtk.h"
+#include "iop/iop_api.h"
+
+#include <gtk/gtk.h>
+#include <stdlib.h>
+#include <string.h>
+#include <math.h>
+#include <float.h>
+
+//----------------------------------------------------------------------
+// implement the module api
+//----------------------------------------------------------------------
+
+DT_MODULE_INTROSPECTION(1, dt_iop_hazeremoval_params_t)
+
+typedef float rgb_pixel[3];
+
+typedef struct dt_iop_hazeremoval_params_t
+{
+  float strength;
+  float distance;
+} dt_iop_hazeremoval_params_t;
+
+// types  dt_iop_hazeremoval_params_t and dt_iop_hazeremoval_data_t are
+// equal, thus no commit_params function needs to be implemented
+typedef dt_iop_hazeremoval_params_t dt_iop_hazeremoval_data_t;
+
+typedef struct dt_iop_hazeremoval_gui_data_t
+{
+  GtkWidget *strength;
+  GtkWidget *distance;
+  rgb_pixel A0;
+  float distance_max;
+  uint64_t hash;
+  dt_pthread_mutex_t lock;
+} dt_iop_hazeremoval_gui_data_t;
+
+typedef struct dt_iop_hazeremoval_global_data_t
+{
+} dt_iop_hazeremoval_global_data_t;
+
+const char *name()
+{
+  return _("haze removal");
+}
+
+int flags()
+{
+  return IOP_FLAGS_INCLUDE_IN_STYLES | IOP_FLAGS_SUPPORTS_BLENDING;
+}
+
+int groups()
+{
+  return IOP_GROUP_CORRECT;
+}
+
+void init_pipe(struct dt_iop_module_t *self, dt_dev_pixelpipe_t *pipe, dt_dev_pixelpipe_iop_t *piece)
+{
+  piece->data = calloc(1, sizeof(dt_iop_hazeremoval_data_t));
+  self->commit_params(self, self->default_params, pipe, piece);
+}
+
+void cleanup_pipe(struct dt_iop_module_t *self, dt_dev_pixelpipe_t *pipe, dt_dev_pixelpipe_iop_t *piece)
+{
+  free(piece->data);
+  piece->data = NULL;
+}
+
+void init_global(dt_iop_module_so_t *self)
+{
+  self->data = NULL;
+}
+
+void cleanup_global(dt_iop_module_so_t *self)
+{
+}
+
+void init(dt_iop_module_t *self)
+{
+  self->params = calloc(1, sizeof(dt_iop_hazeremoval_params_t));
+  self->default_params = calloc(1, sizeof(dt_iop_hazeremoval_params_t));
+  self->default_enabled = 0;
+  self->priority = 338; // module order created by iop_dependencies.py, do not edit!
+  self->params_size = sizeof(dt_iop_hazeremoval_params_t);
+  self->gui_data = NULL;
+  dt_iop_hazeremoval_params_t tmp
+    = (dt_iop_hazeremoval_params_t){ 0.5f, 0.25f };
+  memcpy(self->params, &tmp, sizeof(dt_iop_hazeremoval_params_t));
+  memcpy(self->default_params, &tmp, sizeof(dt_iop_hazeremoval_params_t));
+}
+
+void cleanup(dt_iop_module_t *self)
+{
+  free(self->params);
+  self->params = NULL;
+}
+
+static void strength_callback(GtkWidget *w, dt_iop_module_t *self)
+{
+  if (self->dt->gui->reset)
+    return;
+  dt_iop_hazeremoval_params_t *p = (dt_iop_hazeremoval_params_t *)self->params;
```

Comment:
I can remove this cast, no problem. I my opinion, however, this is not a brilliant idea. `self->parameter` is of type `dt_iop_params_t *`. Removing the cast in the code above is only valid because `dt_iop_params_t` is actually a typedef for `void`. I think, one should treat `dt_iop_params_t` as an opaque type and not rely on some implementation details of this type. Otherwise, the additional layer of abstraction that is gained by introducing the type `dt_iop_params_t` is immediately lost. A possible future change in the implementation of `dt_iop_params_t` will break the code if the explicit pointer cast is removed.